In [1]:
from utils import max_features, maxlen, embed_size_fastText, embed_size_glove, embed_size_glove_twitter
from utils import earlystop, batch_size, epochs, run_5fold, write_prediction
from model import get_textCNN
import pickle
import tensorflow as tf
import pandas as pd

/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #3
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into whic

In [2]:
fastText = pickle.load(open("input/fastText.300d.pickle", "rb"))
glove = pickle.load(open("input/glove.300d.pickle", "rb"))
glove_twitter = pickle.load(open("input/glove.twitter.200d.pickle", "rb"))

In [3]:
x_train = pickle.load(open("input/x_train.pickle", "rb"))
y_train = pickle.load(open("input/y_train.pickle", "rb"))
x_test = pickle.load(open("input/x_test.pickle", "rb"))
submission = pd.read_csv("input/sample_submission.csv")

## check different embedding

In [4]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "num_filter": 300,
    "reg": 0.0,
    "dropout": 0.2
}

In [5]:
embeddings = [(fastText, embed_size_fastText, "fastText"), (glove, embed_size_glove, "glove"), (glove_twitter, embed_size_glove_twitter, "glove_twitter")]

In [6]:
predictions = []
for embed in embeddings:
    print ("\n********************{}********************".format(embed[2]))
    params["embedding_matrix"] = embed[0]
    params["embed_size"] = embed[1]
    y_preds = run_5fold(x_train, y_train, x_test, get_textCNN, **params)
    predictions.append(y_preds)


********************fastText********************
Instructions for updating:
Please switch to tf.metrics.auc. Note that the order of the labels and predictions arguments has been switched.
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 188s - loss: 0.0832 - acc: 0.9782 - auc_roc: 0.9017 - val_loss: 0.0570 - val_acc: 0.9792 - val_auc_roc: 0.9507
Epoch 2/50
 - 185s - loss: 0.0402 - acc: 0.9851 - auc_roc: 0.9632 - val_loss: 0.0467 - val_acc: 0.9831 - val_auc_roc: 0.9699
Epoch 3/50
 - 185s - loss: 0.0326 - acc: 0.9877 - auc_roc: 0.9743 - val_loss: 0.0534 - val_acc: 0.9813 - val_auc_roc: 0.9770
Epoch 4/50
 - 185s - loss: 0.0265 - acc: 0.9901 - auc_roc: 0.9793 - val_loss: 0.0582 - val_acc: 0.9822 - val_auc_roc: 0.9805
Epoch 5/50
 - 185s - loss: 0.0205 - acc: 0.9924 - auc_roc: 0.9818 - val_loss: 0.0648 - val_acc: 0.9817 - val_auc_roc: 0.9826
Epoch 6/50
 - 185s - loss: 0.0168 - acc: 0.9940 - auc_roc: 0.9833 - val_loss: 0.0698 - val_acc: 0.9814 - val_auc_roc: 0.9838
Epoch 7/50

Epoch 10/50
 - 185s - loss: 0.0099 - acc: 0.9969 - auc_roc: 0.9849 - val_loss: 0.0903 - val_acc: 0.9814 - val_auc_roc: 0.9849
Epoch 00010: early stopping
val loss: 0.09031321226190847
val acc: 0.9813509013947593
val auc roc: 0.9848663507864844
loss: 0.09427500354623068 (+/- 0.005256365888987164)
acc: 0.9812089704577811 (+/- 0.00031764766811399794)
roc_auc: 0.9848968426160465 (+/- 0.00024680293672202743)

********************glove********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 186s - loss: 0.1020 - acc: 0.9771 - auc_roc: 0.8785 - val_loss: 0.0509 - val_acc: 0.9819 - val_auc_roc: 0.9425
Epoch 2/50
 - 185s - loss: 0.0428 - acc: 0.9844 - auc_roc: 0.9569 - val_loss: 0.0475 - val_acc: 0.9825 - val_auc_roc: 0.9651
Epoch 3/50
 - 185s - loss: 0.0350 - acc: 0.9870 - auc_roc: 0.9704 - val_loss: 0.0499 - val_acc: 0.9828 - val_auc_roc: 0.9735
Epoch 4/50
 - 185s - loss: 0.0279 - acc: 0.9895 - auc_roc: 0.9763 - val_loss: 0.0565 - val_acc: 0.9814 - val_auc_roc: 0.

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 186s - loss: 0.0880 - acc: 0.9782 - auc_roc: 0.8930 - val_loss: 0.0589 - val_acc: 0.9820 - val_auc_roc: 0.9445
Epoch 2/50
 - 185s - loss: 0.0447 - acc: 0.9842 - auc_roc: 0.9564 - val_loss: 0.0510 - val_acc: 0.9820 - val_auc_roc: 0.9636
Epoch 3/50
 - 185s - loss: 0.0352 - acc: 0.9869 - auc_roc: 0.9686 - val_loss: 0.0521 - val_acc: 0.9822 - val_auc_roc: 0.9719
Epoch 4/50
 - 185s - loss: 0.0289 - acc: 0.9891 - auc_roc: 0.9746 - val_loss: 0.0544 - val_acc: 0.9823 - val_auc_roc: 0.9763
Epoch 5/50
 - 185s - loss: 0.0231 - acc: 0.9915 - auc_roc: 0.9779 - val_loss: 0.0652 - val_acc: 0.9810 - val_auc_roc: 0.9790
Epoch 6/50
 - 185s - loss: 0.0187 - acc: 0.9932 - auc_roc: 0.9800 - val_loss: 0.0652 - val_acc: 0.9817 - val_auc_roc: 0.9809
Epoch 7/50
 - 185s - loss: 0.0158 - acc: 0.9945 - auc_roc: 0.9816 - val_loss: 0.0789 - val_acc: 0.9787 - val_auc_roc: 0.9823
Epoch 8/50
 - 185s - loss: 0.0133 - acc: 0.9956 - auc_roc: 0.9829 - val_lo

Epoch 8/50
 - 148s - loss: 0.0159 - acc: 0.9944 - auc_roc: 0.9811 - val_loss: 0.0768 - val_acc: 0.9820 - val_auc_roc: 0.9813
Epoch 9/50
 - 148s - loss: 0.0150 - acc: 0.9948 - auc_roc: 0.9815 - val_loss: 0.0799 - val_acc: 0.9826 - val_auc_roc: 0.9817
Epoch 10/50
 - 148s - loss: 0.0121 - acc: 0.9959 - auc_roc: 0.9819 - val_loss: 0.0795 - val_acc: 0.9812 - val_auc_roc: 0.9823
Epoch 11/50
 - 148s - loss: 0.0119 - acc: 0.9960 - auc_roc: 0.9826 - val_loss: 0.0796 - val_acc: 0.9820 - val_auc_roc: 0.9828
Epoch 12/50
 - 148s - loss: 0.0098 - acc: 0.9968 - auc_roc: 0.9830 - val_loss: 0.0866 - val_acc: 0.9822 - val_auc_roc: 0.9831
Epoch 13/50
 - 148s - loss: 0.0101 - acc: 0.9968 - auc_roc: 0.9832 - val_loss: 0.0911 - val_acc: 0.9812 - val_auc_roc: 0.9833
Epoch 14/50
 - 148s - loss: 0.0092 - acc: 0.9971 - auc_roc: 0.9834 - val_loss: 0.0971 - val_acc: 0.9822 - val_auc_roc: 0.9833
Epoch 15/50
 - 148s - loss: 0.0083 - acc: 0.9974 - auc_roc: 0.9833 - val_loss: 0.0976 - val_acc: 0.9810 - val_auc_roc: 0

##### result: fastText
- fastText: 

loss: 0.09427500354623068 (+/- 0.005256365888987164)

acc: 0.9812089704577811 (+/- 0.00031764766811399794)

roc_auc: 0.9848968426160465 (+/- 0.00024680293672202743)

- glove: 

loss: 0.10083205023253158 (+/- 0.004831297009038677)

acc: 0.9814679993404226 (+/- 0.0005678382241098716)

roc_auc: 0.9840923237314264 (+/- 0.0004596957303558498)

- glove.twitter: 

loss: 0.10348110079473816 (+/- 0.0070714678773855)

acc: 0.9811316798212382 (+/- 0.00038204777063857964)

roc_auc: 0.9831429467120568 (+/- 0.0010351572143133417)


## check num_filter
(suggest values: 100~600: test 100, 300, 500)

In [4]:
filters = [50, 100, 500]  # for the result of 300, refer to the previous log

In [5]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "reg": 0.0,
    "dropout": 0.2,
    "embedding_matrix": fastText,
    "embed_size": embed_size_fastText
}

In [10]:
predictions = {}
for f in filters:
    print ("\n********************f = {}********************".format(f))
    params["num_filter"] = f
    y_preds = run_5fold(x_train, y_train, x_test, get_textCNN, **params)
    predictions[f] = y_preds


********************f = 50********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 68s - loss: 0.0749 - acc: 0.9771 - auc_roc: 0.9160 - val_loss: 0.0480 - val_acc: 0.9819 - val_auc_roc: 0.9627
Epoch 2/50
 - 67s - loss: 0.0416 - acc: 0.9844 - auc_roc: 0.9714 - val_loss: 0.0468 - val_acc: 0.9822 - val_auc_roc: 0.9762
Epoch 3/50
 - 67s - loss: 0.0335 - acc: 0.9871 - auc_roc: 0.9795 - val_loss: 0.0492 - val_acc: 0.9817 - val_auc_roc: 0.9817
Epoch 4/50
 - 67s - loss: 0.0265 - acc: 0.9898 - auc_roc: 0.9836 - val_loss: 0.0504 - val_acc: 0.9823 - val_auc_roc: 0.9848
Epoch 5/50
 - 67s - loss: 0.0203 - acc: 0.9924 - auc_roc: 0.9860 - val_loss: 0.0565 - val_acc: 0.9818 - val_auc_roc: 0.9867
Epoch 6/50
 - 67s - loss: 0.0161 - acc: 0.9941 - auc_roc: 0.9874 - val_loss: 0.0623 - val_acc: 0.9811 - val_auc_roc: 0.9878
Epoch 7/50
 - 67s - loss: 0.0125 - acc: 0.9956 - auc_roc: 0.9882 - val_loss: 0.0676 - val_acc: 0.9814 - val_auc_roc: 0.9884
Epoch 8/50
 - 67s - loss: 0.0107 

Epoch 2/50
 - 87s - loss: 0.0402 - acc: 0.9847 - auc_roc: 0.9719 - val_loss: 0.0457 - val_acc: 0.9829 - val_auc_roc: 0.9767
Epoch 3/50
 - 87s - loss: 0.0325 - acc: 0.9876 - auc_roc: 0.9799 - val_loss: 0.0507 - val_acc: 0.9819 - val_auc_roc: 0.9819
Epoch 4/50
 - 87s - loss: 0.0260 - acc: 0.9902 - auc_roc: 0.9836 - val_loss: 0.0531 - val_acc: 0.9816 - val_auc_roc: 0.9848
Epoch 5/50
 - 87s - loss: 0.0197 - acc: 0.9928 - auc_roc: 0.9860 - val_loss: 0.0620 - val_acc: 0.9802 - val_auc_roc: 0.9868
Epoch 6/50
 - 87s - loss: 0.0154 - acc: 0.9944 - auc_roc: 0.9876 - val_loss: 0.0660 - val_acc: 0.9814 - val_auc_roc: 0.9879
Epoch 7/50
 - 87s - loss: 0.0122 - acc: 0.9958 - auc_roc: 0.9882 - val_loss: 0.0729 - val_acc: 0.9811 - val_auc_roc: 0.9883
Epoch 8/50
 - 87s - loss: 0.0105 - acc: 0.9964 - auc_roc: 0.9884 - val_loss: 0.0778 - val_acc: 0.9799 - val_auc_roc: 0.9886
Epoch 9/50
 - 87s - loss: 0.0093 - acc: 0.9969 - auc_roc: 0.9888 - val_loss: 0.0829 - val_acc: 0.9818 - val_auc_roc: 0.9886
Epoch 10

Epoch 2/50
 - 258s - loss: 0.0473 - acc: 0.9845 - auc_roc: 0.9549 - val_loss: 0.0510 - val_acc: 0.9825 - val_auc_roc: 0.9616
Epoch 3/50
 - 258s - loss: 0.0358 - acc: 0.9870 - auc_roc: 0.9668 - val_loss: 0.0515 - val_acc: 0.9824 - val_auc_roc: 0.9703
Epoch 4/50
 - 258s - loss: 0.0277 - acc: 0.9898 - auc_roc: 0.9733 - val_loss: 0.0645 - val_acc: 0.9818 - val_auc_roc: 0.9748
Epoch 5/50
 - 258s - loss: 0.0229 - acc: 0.9916 - auc_roc: 0.9761 - val_loss: 0.0687 - val_acc: 0.9818 - val_auc_roc: 0.9771
Epoch 6/50
 - 258s - loss: 0.0184 - acc: 0.9935 - auc_roc: 0.9780 - val_loss: 0.0785 - val_acc: 0.9814 - val_auc_roc: 0.9785
Epoch 7/50
 - 258s - loss: 0.0156 - acc: 0.9946 - auc_roc: 0.9791 - val_loss: 0.0811 - val_acc: 0.9815 - val_auc_roc: 0.9795
Epoch 8/50
 - 258s - loss: 0.0135 - acc: 0.9955 - auc_roc: 0.9800 - val_loss: 0.0824 - val_acc: 0.9810 - val_auc_roc: 0.9805
Epoch 9/50
 - 258s - loss: 0.0120 - acc: 0.9962 - auc_roc: 0.9809 - val_loss: 0.0870 - val_acc: 0.9810 - val_auc_roc: 0.9812


 - 258s - loss: 0.0137 - acc: 0.9954 - auc_roc: 0.9807 - val_loss: 0.0827 - val_acc: 0.9808 - val_auc_roc: 0.9811
Epoch 9/50
 - 258s - loss: 0.0126 - acc: 0.9960 - auc_roc: 0.9815 - val_loss: 0.1009 - val_acc: 0.9816 - val_auc_roc: 0.9813
Epoch 10/50
 - 258s - loss: 0.0110 - acc: 0.9965 - auc_roc: 0.9811 - val_loss: 0.0941 - val_acc: 0.9818 - val_auc_roc: 0.9812
Epoch 11/50
 - 258s - loss: 0.0098 - acc: 0.9970 - auc_roc: 0.9813 - val_loss: 0.0956 - val_acc: 0.9815 - val_auc_roc: 0.9814
Epoch 12/50
 - 258s - loss: 0.0090 - acc: 0.9972 - auc_roc: 0.9816 - val_loss: 0.1054 - val_acc: 0.9812 - val_auc_roc: 0.9816
Epoch 13/50
 - 258s - loss: 0.0089 - acc: 0.9974 - auc_roc: 0.9817 - val_loss: 0.1133 - val_acc: 0.9819 - val_auc_roc: 0.9815
Epoch 14/50
 - 258s - loss: 0.0084 - acc: 0.9976 - auc_roc: 0.9814 - val_loss: 0.1103 - val_acc: 0.9811 - val_auc_roc: 0.9813
Epoch 00014: early stopping
val loss: 0.11034916334908028
val acc: 0.9811158985200215
val auc roc: 0.9813264284081915
loss: 0.11403

###### result: 50

- f = 50

loss: 0.08239762419408467 (+/- 0.0018483503651972517)

acc: 0.9811922631198764 (+/- 0.00044086702309639936)

roc_auc: 0.9882647025436222 (+/- 0.00019733618150735181)

- f = 100

loss: 0.08705983650200938 (+/- 0.003124276775068534)

acc: 0.9811692809088642 (+/- 0.0002826450893100738)

roc_auc: 0.987828371328319 (+/- 0.0004778370095814332)

- f = 300

loss: 0.09427500354623068 (+/- 0.005256365888987164)

acc: 0.9812089704577811 (+/- 0.00031764766811399794)

roc_auc: 0.9848968426160465 (+/- 0.00024680293672202743)

- f = 500

loss: 0.11403924442061053 (+/- 0.013918692019939043)

acc: 0.9807149444437446 (+/- 0.000641540160284363)

roc_auc: 0.982061221822742 (+/- 0.0009086290070467134)

In [11]:
optimal_pred = predictions[50]

In [15]:
pickle.dump(optimal_pred, open("curr_opt.pickle", "wb"))

## check kernel_size
(suggest value: [3, 5, 7, 9])

In [4]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "dropout": 0.2,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": fastText,
    "embed_size": embed_size_fastText
}

In [5]:
kernels = [5, 7, 9]  # for 3, see result above

In [6]:
predictions = {}
for k in kernels:
    print ("\n********************k = {}********************".format(k))
    params["kernel_size"] = k
    y_preds = run_5fold(x_train, y_train, x_test, get_textCNN, **params)
    predictions[k] = y_preds


********************k = 5********************
Instructions for updating:
Please switch to tf.metrics.auc. Note that the order of the labels and predictions arguments has been switched.
Epoch 00010: early stopping
val loss: 0.0860074702092945
val acc: 0.98142982078312
val auc roc: 0.9883746058780647

Epoch 00011: early stopping
val loss: 0.08222856482693655
val acc: 0.980823440308812
val auc roc: 0.9884522493459187

Epoch 00013: early stopping
val loss: 0.08516958197152143
val acc: 0.981700794921077
val auc roc: 0.9884031944954135

Epoch 00013: early stopping
val loss: 0.08697717188117259
val acc: 0.9814971244640724
val auc roc: 0.9879353151043802

Epoch 00013: early stopping
val loss: 0.0861990955199687
val acc: 0.9803116538773102
val auc roc: 0.9882568011212455

loss: 0.08531637688177876 (+/- 0.0016474935048893234)
acc: 0.9811525668708784 (+/- 0.0005121723574178807)
roc_auc: 0.9882844331890045 (+/- 0.00018603472388464064)



********************k = 7********************
Epoch 00011: 

##### result: 9

- k = 3

loss: 0.08239762419408467 (+/- 0.0018483503651972517)

acc: 0.9811922631198764 (+/- 0.00044086702309639936)

roc_auc: 0.9882647025436222 (+/- 0.00019733618150735181)


- k = 5

loss: 0.08531637688177876 (+/- 0.0016474935048893234)

acc: 0.9811525668708784 (+/- 0.0005121723574178807)

roc_auc: 0.9882844331890045 (+/- 0.00018603472388464064)

- k = 7

loss: 0.08634472576870844 (+/- 0.0012858760631471353)

acc: 0.9813249081564093 (+/- 0.00031133671420321204)

roc_auc: 0.9881019192218421 (+/- 0.00034188073297824247)

- k = 9

loss: 0.0830957416998463 (+/- 0.0023434830099651714)

acc: 0.981463826513648 (+/- 0.0004240916565512265)

roc_auc: 0.9884804397433369 (+/- 0.0005250787379544356)

In [7]:
optimal_pred = predictions[9]

## check drop out
(suggest value 0.1~0.5: test 0.1, 0.2, 0.5)

In [8]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "kernel_size": 9,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": fastText,
    "embed_size": embed_size_fastText
}

In [9]:
dropouts = [0.1, 0.5] # 0.2 see previous result

In [10]:
predictions = {}
for d in dropouts:
    print ("\n********************d = {}********************".format(d))
    params["dropout"] = d
    y_preds = run_5fold(x_train, y_train, x_test, get_textCNN, **params)
    predictions[d] = y_preds


********************d = 0.1********************
Epoch 00009: early stopping
val loss: 0.08166127521500922
val acc: 0.9799989365026082
val auc roc: 0.9882197980930789

Epoch 00012: early stopping
val loss: 0.08561740595651963
val acc: 0.9803273196491433
val auc roc: 0.9880629033644023

Epoch 00013: early stopping
val loss: 0.08548564512833337
val acc: 0.9807659943554868
val auc roc: 0.9883104757444746

Epoch 00010: early stopping
val loss: 0.08396109096351524
val acc: 0.980781666122523
val auc roc: 0.9880370349336285

Epoch 00009: early stopping
val loss: 0.09083764221568959
val acc: 0.9812569013923573
val auc roc: 0.9871675177625138

loss: 0.08551261189581341 (+/- 0.0030201876875974244)
acc: 0.9806261636044237 (+/- 0.0004299500845598379)
roc_auc: 0.9879595459796198 (+/- 0.0004086266868660403)



********************d = 0.5********************
Epoch 00020: early stopping
val loss: 0.08432116117109545
val acc: 0.9815081496364212
val auc roc: 0.9888406693683267

Epoch 00019: early stoppi

##### result

- d = 0.1

loss: 0.08551261189581341 (+/- 0.0030201876875974244)

acc: 0.9806261636044237 (+/- 0.0004299500845598379)

roc_auc: 0.9879595459796198 (+/- 0.0004086266868660403)

- d = 0.2

loss: 0.0830957416998463 (+/- 0.0023434830099651714)

acc: 0.981463826513648 (+/- 0.0004240916565512265)

roc_auc: 0.9884804397433369 (+/- 0.0005250787379544356)

- d = 0.5

loss: 0.08056822570324007 (+/- 0.0040002786631913095)

acc: 0.9816278017943952 (+/- 0.00024987610223536796)

roc_auc: 0.988724607426645 (+/- 0.00030711458676242965)

In [12]:
optimal_pred = predictions[0.5]

#### tuned params

In [14]:
params_textCNN = {
    "maxlen": maxlen,
    "max_features": max_features,
    "kernel_size": 9,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": fastText,
    "embed_size": embed_size_fastText,
    "dropout": 0.5
}

### write result

In [13]:
write_prediction(optimal_pred, submission, "output/textCNN.csv")